### PR TITLE
Rewrite in Rust 🦀

### PR DESCRIPTION
**Pull request by dbrgn**
With two payload format versions, and now with different brands (our own and the ones by Dragino), things get quite complex with the Python script. Testing is hard. I'd prefer compile time checks.

There's only one solution: RIIR!

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+anyhow = "1"
+clap = { version = "3.1.18", features = ["derive"] }
+drogue-ttn = "0.5"
+env_logger = "0.9"
+log = "0.4"
+paho-mqtt = "0.11.1"
+reqwest = { version = "0.11", features = ["blocking"] }
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+toml = "0.5"
```
Comment (rnestler) [inline Cargo.toml:17]: No dependency to the firmware message encode / decode crate? Would be nice to implement it only once.

Comment (dbrgn) [reply Cargo.toml:17]: Yeah, that part isn't implemented yet.

It's very nice though that there's the `dragino-ttn` crate for deserializing TTN payloads!

Comment (dbrgn) [reply Cargo.toml:17]: Since the firmware decoding crate is part of the firmware repo: Should we publish it to crates.io?

Comment (dbrgn) [reply Cargo.toml:18]: Ping @rnestler, what do you think? (An alternative would be submodules...)

Comment (rnestler) [reply Cargo.toml:17]: Just a git dependency should work as well, no? There is not much value in putting it on crates.io IMO.

Comment (dbrgn) [reply Cargo.toml:17]: Ah, good point!

Comment (dbrgn) [reply Cargo.toml:17]: Uhm, it seems the crate currently handles only encoding, but not decoding 😅

**Comment by dbrgn** [comment]
Auto-reconnect doesn't seem to work yet:

    [2022-06-14T08:22:07Z WARN  paho_mqtt::async_client] Connection lost. Context: 0x563ac635cf50
    [2022-06-14T08:22:07Z WARN  ttn_relay] Connection lost. Waiting to retry connection

**Comment by dbrgn** [comment]
Another parse error:

```json
{
    "end_device_ids": {
        "device_id": "dragino-4bca",
        "application_ids": {
            "application_id": "gfroerli-live"
        },
        "dev_eui": "A840414E81844BCA",
        "join_eui": "A840410000000101",
        "dev_addr": "260BE33B"
    },
    "correlation_ids": [
        "as:up:01G9J2B8K096XKGHJ3484QN96P",
        "gs:conn:01G88BAPY9JMZHH09DV9HZDS1V",
        "gs:up:host:01G88BAQG2Y9R78PNA379BB5DR",
        "gs:uplink:01G9J2B8CHQQ9BRND893P5J732",
        "ns:uplink:01G9J2B8CJFQMWSWAK8VJ1NKK3",
        "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01G9J2B8CJ6WA56KKAJ8NF1CVJ",
        "rpc:/ttn.lorawan.v3.NsAs/HandleUplink:01G9J2B8K0KWDWW3SAXK4BJP3W"
    ],
    "received_at": "2022-08-03T14:39:24.256954911Z",
    "uplink_message": {
        "session_key_id": "AYDxsEf/RwoWuMQy5qmVAw==",
        "f_port": 3,
        "f_cnt": 6906,
        "frm_payload": "DkoBIAAADP////8=",
        "rx_metadata": [
            {
                "gateway_ids": {
                    "gateway_id": "coredump01",
                    "eui": "B827EBFFFE4A5213"
                },
                "time": "2022-08-03T14:39:24.030283Z",
                "timestamp": 1200514995,
                "rssi": -114,
                "channel_rssi": -114,
                "location": {
                    "latitude": 47.227625139237624,
                    "longitude": 8.820889592170717,
                    "altitude": 419,
                    "source": "SOURCE_REGISTRY"
                },
                "uplink_token": "ChgKFgoKY29yZWR1bXAwMRIIuCfr//5KUhMQs8+5vAQaCwickaqXBhCdt7cXILimlKL40L4CKgsInJGqlwYQ+Km4Dg==",
                "channel_index": 1
            }
        ],
        "settings": {
            "data_rate": {
                "lora": {
                    "bandwidth": 125000,
                    "spreading_factor": 7
                }
            },
            "coding_rate": "4/5",
            "frequency": "868300000",
            "timestamp": 1200514995,
            "time": "2022-08-03T14:39:24.030283Z"
        },
        "received_at": "2022-08-03T14:39:24.050221145Z",
        "consumed_airtime": "0.061696s",
        "locations": {
            "user": {
                "latitude": 47.23333059911948,
                "longitude": 8.817209601402284,
                "altitude": 406,
                "source": "SOURCE_REGISTRY"
            }
        },
        "version_ids": {
            "brand_id": "dragino",
            "model_id": "lsn50v2-d20",
            "hardware_version": "_unknown_hw_version_",
            "firmware_version": "1.1",
            "band_id": "EU_863_870"
        },
        "network_ids": {
            "net_id": "000013",
            "tenant_id": "ttn",
            "cluster_id": "eu1",
            "cluster_address": "eu1.cloud.thethings.network"
        }
    }
}
```

**Comment by dbrgn** [comment]
TODO: Send gateway information to InfluxDB

**Comment by dbrgn** [comment]
Reconnect doesn't seem to work:

```
[2022-08-22T14:36:36Z WARN  paho_mqtt::async_client] Connection lost. Context: 0x564eef0515b0
[2022-08-22T14:36:36Z WARN  ttn_relay] Connection lost. Waiting to retry connection
[2022-08-22T14:36:41Z INFO  ttn_relay] Reconnection attempt 1
[2022-08-22T14:41:46Z INFO  ttn_relay] Reconnection attempt 2
[2022-08-22T14:46:51Z INFO  ttn_relay] Reconnection attempt 3
[2022-08-22T14:51:56Z INFO  ttn_relay] Reconnection attempt 4
[2022-08-22T14:57:01Z INFO  ttn_relay] Reconnection attempt 5
[2022-08-22T15:02:06Z INFO  ttn_relay] Reconnection attempt 6
[2022-08-22T15:07:11Z INFO  ttn_relay] Reconnection attempt 7
[2022-08-22T15:12:16Z INFO  ttn_relay] Reconnection attempt 8
[2022-08-22T15:17:21Z INFO  ttn_relay] Reconnection attempt 9
[2022-08-22T15:22:26Z INFO  ttn_relay] Reconnection attempt 10
[2022-08-22T15:27:31Z INFO  ttn_relay] Reconnection attempt 11
```

**Comment by dbrgn** [comment]
Alright, the Rust handler for the Dragino sensors is now feature-complete. Only remaining issue is handling / decoding for our own format.

File: Cargo.toml (hunk 1 above)
Comment (dbrgn) [reply Cargo.toml:17]: Oh: https://github.com/gfroerli/firmware/pull/131

**Pull request merged**